### ISSUE TITLE
CEMExplainer scikit-learn compatibility & "predict_long" method error

### ISSUE THREAD

**Issue by josefinarcasanova** [Trusted-AI/AIX360#152]
Hello! I've got a problem you might be able to help me with.

I'm trying to use CEM to explain `scikit-learn` binary classification models. I'm using `RandomForestClassifier` & `KNeighborsClassifier` to be more exact. 

Here's a snipet using K-Neighbors:

```python
from sklearn.neighbors import KNeighborsClassifier

model_kneighbors = KNeighborsClassifier()

fit_kneighbors = model_kneighbors.fit(X_train, y_train_)
y_pred_kneighbors = model_kneighbors.predict(X_test)

cem_kneighbors_clas = CEMExplainer(model_kneighbors)

arg_mode = 'PN'             # Find pertinent negatives
arg_max_iter = 1000         # Maximum number of iterations to search for the optimal PN for given parameter settings
arg_init_const = 10.0       # Initial coefficient value for main loss term that encourages class change
arg_b = 9                   # No. of updates to the coefficient of the main loss term
arg_kappa = 0.2             # Minimum confidence gap between the PNs (changed) class probability and original class' probability
arg_beta = 1e-1             # Controls sparsity of the solution (L1 loss)
arg_gamma = 100             # Controls how much to adhere to a (optionally trained) auto-encoder
my_AE_model = None          # Pointer to an auto-encoder
arg_alpha = 0.01            # Penalizes L2 norm of the solution
arg_threshold = 1.          # Automatically turn off features <= arg_threshold if arg_threshold < 1
arg_offset = 0.5            # the model assumes classifier trained on data normalized
                            # in [-arg_offset, arg_offset] range, where arg_offset is 0 or 0.5

(adv_pn, delta_pn, info_pn) = cem_kneighbors_clas.explain_instance(
    input_X=X_to_explain_clas,      # input_X (numpy.ndarray) – input instance to be explained
    arg_mode=arg_mode,              # arg_mode (str) – ‘PP’ or ‘PN’
    AE_model=my_AE_model,           # AE_model – Auto-encoder model
    arg_kappa=arg_kappa,            # arg_kappa (double) – Confidence gap between desired class and other classes
    arg_b=arg_b,                    # arg_b (double) – Number of different weightings of loss function to try
    arg_max_iter=arg_max_iter,      # arg_max_iter (int) – For each weighting of loss function number of iterations to search
    arg_init_const=arg_init_const,  # arg_init_const (double) – Initial weighting of loss function
    arg_beta=arg_beta,              # arg_beta (double) – Weighting of L1 loss
    arg_gamma=arg_gamma             # arg_gamma (double) – Weighting of auto-encoder                  
)
```

When I try to run it I get: `'KNeighborsClassifier' object has no attribute 'predict_long'`. I checked out the [CEM implementation](https://github.com/Trusted-AI/AIX360/blob/master/aix360/algorithms/contrastive/CEM.py) and found the `predict_long` call. I also tried the [HELOC tutorial example](https://github.com/Trusted-AI/AIX360/blob/master/examples/tutorials/HELOC.ipynb) and it all works well, but with a classifier named `KerasClassifier`, that I think has been deprecated? I can't find documentation for it anywhere.

Based on this, I've got a few questions:

1. What estimators are compatible with CEM? The other classes are compatible with `scikit-learn`, but perhaps CEM is for NNs or TensorFlow only? I skimmed the paper but didn't quite catch that. I know there's a variation that deals with images, but "regular" CEM should work for tabular data, correct? Since `ProtoDash` is compatible with these estimators, I assumed CEM was too.
2. Perhaps I simply need to downgrade the package? I didn't install anything in particular in my Jupyter Notebook, just `aix360` (didn't specify a version either, just ran `pip install aix360`).
3. If I were to "tweak" the implementation and change `predict_long` for the usual `scikit-learn` `predict` method, would it work? I don't fully understand the code, to be honest.
4. If `scikit-learn` estimators aren't supported for CEM at the moment, is there an implementation planned for the future?

Superb work with the library, by the way. I'm a huge fan of IBM Research and the amazing work you all do.

Regards from a fellow IBMer!

Kindly,
Josefina

**Comment by sadhamanus** [comment]
Hi Josefina. Thanks for your comments. The main issue in your case is I think you are applying CEM to non-differentiable models. CEM is a white box method which requires you to backpropagate through the model unlike Protodash etc. Currently, variants of this algorithm applicable to black box classifiers are available through IBM product and paid research offerings.

**Comment by vijay-arya** [comment]
@josefinarcasanova Josefina, based on the issue @sadhamanus pointed, one possible way to overcome your problem (with a bit of tradeoff) could be the following. 

If you are keen about continuing to use a scikit-learn classifier (random-forest, KNN), then after training these model, use them to relabel your source train/test dataset. Then train a surrogate neural network (NN) model on the relabelled data, then explain any data instance via CEM. Most likely the NN will approximate for your source scikit-learn model well. 

If you are not keen about using a scikit-learn classifier and are willing to switch to a NN model, then your task is similar to the one in HELOC tutorial, just train a NN on your source dataset. Take a look at the NN training part in HELOC tutorial as you need to output logits (no softmax) to apply CEM later. Hope this helps.

**Comment by josefinarcasanova** [comment]
Hi @vijay-arya, @sadhamanus, how are you?

Thank you both for your answers. They were very useful. I will talk to the rest of my Team to decide what to do, but now everything is much clearer.

I will close the issue so it isn't left lying around.

Thanks again!

Regards,
Josefina
